src/main/java/io/vertx/core/impl/DeploymentManager.java

```diff
@@ -79,13 +99,47 @@ public void deployVerticle(Verticle verticle, DeploymentOptions options,
       throw new IllegalArgumentException("Can't specify isolatedClasses for already created verticle");
     }
     ContextImpl currentContext = vertx.getOrCreateContext();
-    doDeploy("java:" + verticle.getClass().getName(), generateDeploymentID(), options, currentContext, currentContext, completionHandler,
-        getCurrentClassLoader(), verticle);
+    ClassLoader cl = getClassLoader(options, currentContext);
+    currentContext.<Set<Verticle>>executeBlocking(fut -> {
+      int nbInstances = options.getInstances();
+      IdentityHashMap<Class<? extends Verticle>, Set<Verticle>> verticles = new IdentityHashMap<>();
+      for (int i = 0; i < nbInstances; i++) {
+        Verticle verticle = verticleSupplier.get();
+        if (verticle == null) {
+          throw new RuntimeException("Supplied verticle is null");
+        }
+        verticles.compute(verticle.getClass(), (k, v) -> {
+          if (v == null) v = new HashSet<>();
+          v.add(verticle);
+          return v;
+        });
+      }
+      if (verticles.size() != 1) {
+        throw new RuntimeException("Supplied verticles are not of the same class");
```

Comment:
I think your approach is right. Let's start with a conservative approach and if we see that people need a bit more, let's rediscuss it. So 👍